### PR TITLE
Implement excessive sighashing protection policy with tight sighash estimation

### PR DESCRIPTION
**Pull request by jl2012**
This implements a static estimation of sighash size for a transaction. A transaction with more than 90bytes of sighash per weight is non-standard. This is equivalent to 36MB for an 100kB non-segwit transaction, or 360MB for a block in the worst case. All transactions below 100kB with legitimate use of CHECK(MULTI)SIG should remain standard with this limit.

The estimation of sighash is based on the following 3 assumptions:
 a. OP_CODESEPARATOR and FindAndDelete are disabled by SCRIPT_VERIFY_CONST_SCRIPTCODE. This ensures that the scriptCode serialized in SignatureHash is always the same as the original script passing to the EvalScript. (part of this PR)
 b. SignatureHash is performed once only for each SIGHASH type. (#8654)
 c. Only 6 sighash types are allowed: ALL, NONE, SINGLE, and combinations with ANYONECANPAY (already enforced as policy with STRICTENC)

Todo: unit tests

### CODE REVIEW

**Comment by jl2012** [comment]
@TheBlueMatt As you suggested, we could be more aggressive when disabling FindAndDelete. So eventually we may retire this function after a softfork.

File: src/primitives/transaction.cpp [hunk 1]
```diff
@@ -153,3 +153,23 @@ int64_t GetTransactionWeight(const CTransaction& tx)
...
+{
+    int64_t size = ::GetSerializeSize(tx, SER_NETWORK, PROTOCOL_VERSION | SERIALIZE_TRANSACTION_NO_WITNESS);
+    for (unsigned int i = 0; i < tx.vin.size(); i++) {
+        int64_t scriptSigSize = tx.vin[i].scriptSig.size();
+        size -= scriptSigSize;
+        // If the scriptSig size is larger than 252, 2 bytes compactSize encoding is deducted.
+        if (scriptSigSize > 252)
+            size -= 2;
+        /*
+         * // scriptSig larger than 10000 btyes is invalid
```
Comment (instagibbs) [inline src/primitives/transaction.cpp]: dead code

**Comment by jl2012** [comment]
A draft BIP is made for the detailed rationale of this PR: https://github.com/jl2012/bips/blob/sighash/bip-sighash.mediawiki

File: src/primitives/transaction.cpp (hunk 1 above)
Comment (jl2012) [inline src/primitives/transaction.cpp]: removed

**Comment by jl2012** [comment]
Unit tests are completed and related BIP updated

**Comment by laanwj** [comment]
Removing 0.14 tag as discussed in today's meeting

**Comment by TheBlueMatt** [comment]
Strong Concept ACK on at least making CODESEPARATOR and FindAndDelete non-standard. Can we push forward on that independantly to get it done sooner rather than later, then at least the sighash limits can be reviewed separately and are more straight-forward? Any plans to rebase this @jl2012?

**Comment by jl2012** [comment]
@TheBlueMatt : sure, I'll make another PR just for the `SCRIPT_VERIFY_CONST_SCRIPTCODE` policy

**Comment by sipa** [comment]
@jl2012 What's the status here?

**Comment by jl2012** [comment]
@sipa: this requires #8654 and #11423. But #8654 needs to maintain a cache of 256 32-bytes hashes per input which might impact validation. It could be reduced to 6 hashes/input if we softfork away those 250 non-std nHashType.

The alternative is #8756, which does not require #8654 and #11423. But the counting is more conservative (overestimating)

File: src/script/interpreter.h [hunk 2]
```diff
@@ -106,6 +106,10 @@ enum
     // Public keys in segregated witness scripts must be compressed
     //
     SCRIPT_VERIFY_WITNESS_PUBKEYTYPE = (1U << 15),
+
+    // Making OP_CODESEPARATOR and FindAndDelete non-standard in non-segwit scripts
+    //
+    SCRIPT_VERIFY_CONST_SCRIPTCODE = (1U << 16),
```
Comment (MarcoFalke) [inline src/script/interpreter.h]: Does adding a new constant require the tests to be updated as well? See `ValidateCheckInputsForAllFlags`

**Pull request closed without merging**